wolfcrypt/src/port/autosar/crypto.c

```diff
@@ -0,0 +1,457 @@
+            WOLFSSL_MSG("Crypto error with getting an IV");
+            return E_NOT_OK;
+        }
+
+        if (iv != NULL && ivSz < AES_BLOCK_SIZE) {
+            WOLFSSL_MSG("Error IV is too small");
+            return E_NOT_OK;
+        }
+
+        aes = NewAesStruct(job);
+        if (aes == NULL) {
+            WOLFSSL_MSG("Unable to get AES structure for use");
+            return E_NOT_OK;
+        }
+
+        if (wc_AesSetKey(aes, key, keySz, iv, encrypt) != 0) {
+            WOLFSSL_MSG("Crypto error setting up AES key");
+            return E_NOT_OK;
+        }
+        /* ForceZero(key, keySz); Do not keep raw key in memory */
+    }
+
+    if ((job->jobPrimitiveInputOutput.mode & CRYPTO_OPERATIONMODE_UPDATE)
+            != 0) {
+        aes = GetAesStruct(job);
+        if (aes == NULL) {
+            WOLFSSL_MSG("Error finding AES structure");
+            return E_NOT_OK;
+        }
+
+        if (encrypt == AES_ENCRYPTION) {
+            if (wc_AesCbcEncrypt(aes, job->jobPrimitiveInputOutput.outputPtr,
+                    job->jobPrimitiveInputOutput.inputPtr,
+                    job->jobPrimitiveInputOutput.inputLength) != 0) {
+                WOLFSSL_MSG("AES-CBC encrypt error");
+                return E_NOT_OK;
+            }
+        }
+        else {
+            if (wc_AesCbcDecrypt(aes, job->jobPrimitiveInputOutput.outputPtr,
+                    job->jobPrimitiveInputOutput.inputPtr,
+                    job->jobPrimitiveInputOutput.inputLength) != 0) {
+                WOLFSSL_MSG("AES-CBC decrypt error");
+                return E_NOT_OK;
+            }
+        }
+    }
+
+    if ((job->jobPrimitiveInputOutput.mode & CRYPTO_OPERATIONMODE_FINISH)
+            != 0) {
+        FreeAesStruct(job);
+    }
+
+    return ret;
+}
+
+
+/* returns E_OK on success and E_NOT_OK on failure */
+Std_ReturnType wolfSSL_Crypto(Crypto_JobType* job)
+{
+    Std_ReturnType ret = E_OK;
+
+    WOLFSSL_ENTER("wolfSSL_Crypto");
+
+    /* switch on encryption type */
+    switch (job->jobPrimitiveInfo->primitiveInfo->algorithm.mode) {
+        case CRYPTO_ALGOMODE_CBC:
+            ret = wolfSSL_Crypto_CBC(job);
+            break;
+
+        case CRYPTO_ALGOMODE_NOT_SET:
+            WOLFSSL_MSG("Encrypt algo mode not set!");
+            ret = E_NOT_OK;
+            break;
+
+        default:
+            WOLFSSL_MSG("Unsupported encryption mode");
+            ret = E_NOT_OK;
+            break;
+    }
+
+    WOLFSSL_LEAVE("wolfSSL_Crypto", ret);
+    return ret;
+}
+
+
+/* returns E_OK on success */
+Std_ReturnType wolfSSL_Crypto_RNG(Crypto_JobType* job)
+{
+    WC_RNG rng;
+    int ret;
+
+    uint8  *out   = job->jobPrimitiveInputOutput.outputPtr;
+    uint32 *outSz = job->jobPrimitiveInputOutput.outputLengthPtr;
+
+    if (outSz == NULL || out == NULL) {
+        WOLFSSL_MSG("Bad parameter passed into wolfSSL_Crypto_RNG");
+        return E_NOT_OK;
+    }
+
+    ret = wc_InitRng_ex(&rng, NULL, 0);
```

Comment:
We don't want a global?

Reply:
There's no cleanup call Crypto_Cleanup to match the Crypto_Init, is it ok staying alloc'd after program exit with no wc_FreeRng call?

Reply:
I think so. There should be a cleanup if they care about program exit.
I was also worried about threading.